File: src/main/java/com/uiptv/db/SeriesWatchStateDb.java
package com.uiptv.db;

import com.uiptv.model.SeriesWatchState;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;

import static com.uiptv.db.DatabaseUtils.DbTable.SERIES_WATCH_STATE_TABLE;
import static com.uiptv.db.SQLConnection.connect;

public class SeriesWatchStateDb extends BaseDb {
    private static SeriesWatchStateDb instance;

    public SeriesWatchStateDb() {
        super(SERIES_WATCH_STATE_TABLE);
    }

    public static synchronized SeriesWatchStateDb get() {
        if (instance == null) {
            instance = new SeriesWatchStateDb();
        }
        return instance;
    }

    public SeriesWatchState getBySeries(String accountId, String categoryId, String seriesId) {
        List<SeriesWatchState> rows = getAll(" WHERE accountId=? AND mode=? AND categoryId=? AND seriesId=?",
                new String[]{accountId, "series", categoryId, seriesId});
        return rows.isEmpty() ? null : rows.get(0);
    }

    public List<SeriesWatchState> getBySeries(String accountId, String seriesId) {
        return getAll(" WHERE accountId=? AND mode=? AND seriesId=?",
                new String[]{accountId, "series", seriesId});
    }

    public List<SeriesWatchState> getByAccount(String accountId, String categoryId) {
        return getAll(" WHERE accountId=? AND mode=? AND categoryId=?", new String[]{accountId, "series", categoryId});
    }

    public List<SeriesWatchState> getByAccount(String accountId) {
        return getAll(" WHERE accountId=? AND mode=?", new String[]{accountId, "series"});
    }

    public void upsert(SeriesWatchState state) {
        String updateSql = "UPDATE " + SERIES_WATCH_STATE_TABLE.getTableName()
                + " SET episodeId=?, episodeName=?, season=?, episodeNum=?, updatedAt=?, source=?,"
                + " seriesCategorySnapshot=?, seriesChannelSnapshot=?, seriesEpisodeSnapshot=?"
                + " WHERE accountId=? AND mode=? AND categoryId=? AND seriesId=?";
        String insertSql = "INSERT INTO " + SERIES_WATCH_STATE_TABLE.getTableName()
                + " (accountId, mode, categoryId, seriesId, episodeId, episodeName, season, episodeNum, updatedAt, source,"
                + " seriesCategorySnapshot, seriesChannelSnapshot, seriesEpisodeSnapshot)"
                + " VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?)";
        try (Connection conn = connect();
             PreparedStatement update = conn.prepareStatement(updateSql);
             PreparedStatement insert = conn.prepareStatement(insertSql)) {
            update.setString(1, state.getEpisodeId());
            update.setString(2, state.getEpisodeName());
            update.setString(3, state.getSeason());
            update.setInt(4, state.getEpisodeNum());
            update.setLong(5, state.getUpdatedAt());
            update.setString(6, state.getSource());
            update.setString(7, state.getSeriesCategorySnapshot());
            update.setString(8, state.getSeriesChannelSnapshot());
            update.setString(9, state.getSeriesEpisodeSnapshot());
            update.setString(10, state.getAccountId());
            update.setString(11, state.getMode());
            update.setString(12, state.getCategoryId());
            update.setString(13, state.getSeriesId());
            int updated = update.executeUpdate();
            if (updated == 0) {
                insert.setString(1, state.getAccountId());
                insert.setString(2, state.getMode());
                insert.setString(3, state.getCategoryId());
                insert.setString(4, state.getSeriesId());
                insert.setString(5, state.getEpisodeId());
                insert.setString(6, state.getEpisodeName());
                insert.setString(7, state.getSeason());
                insert.setInt(8, state.getEpisodeNum());
                insert.setLong(9, state.getUpdatedAt());
                insert.setString(10, state.getSource());
                insert.setString(11, state.getSeriesCategorySnapshot());
                insert.setString(12, state.getSeriesChannelSnapshot());
                insert.setString(13, state.getSeriesEpisodeSnapshot());
                insert.executeUpdate();
            }
        } catch (SQLException e) {
            throw new RuntimeException("Unable to save series watch state", e);
        }
    }

    public void clear(String accountId, String categoryId, String seriesId) {
        String sql = "DELETE FROM " + SERIES_WATCH_STATE_TABLE.getTableName() + " WHERE accountId=? AND mode=? AND categoryId=? AND seriesId=?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, accountId);
            statement.setString(2, "series");
            statement.setString(3, categoryId);
            statement.setString(4, seriesId);
            statement.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException("Unable to clear series watch state", e);
        }
    }

    public void deleteByAccount(String accountId) {
        String sql = "DELETE FROM " + SERIES_WATCH_STATE_TABLE.getTableName() + " WHERE accountId=?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, accountId);
            statement.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException("Unable to delete watch state by account", e);
        }
    }

    public void clearAllSeries() {
        String sql = "DELETE FROM " + SERIES_WATCH_STATE_TABLE.getTableName() + " WHERE mode=?";
        try (Connection conn = connect(); PreparedStatement statement = conn.prepareStatement(sql)) {
            statement.setString(1, "series");
            statement.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException("Unable to clear all series watch state", e);
        }
    }

    @Override
    SeriesWatchState populate(ResultSet resultSet) {
        SeriesWatchState state = new SeriesWatchState();
        state.setDbId(nullSafeString(resultSet, "id"));
        state.setAccountId(nullSafeString(resultSet, "accountId"));
        state.setMode(nullSafeString(resultSet, "mode"));
        state.setCategoryId(nullSafeString(resultSet, "categoryId"));
        state.setSeriesId(nullSafeString(resultSet, "seriesId"));
        state.setEpisodeId(nullSafeString(resultSet, "episodeId"));
        state.setEpisodeName(nullSafeString(resultSet, "episodeName"));
        state.setSeason(nullSafeString(resultSet, "season"));
        state.setEpisodeNum(safeInteger(resultSet, "episodeNum"));
        try {
            state.setUpdatedAt(resultSet.getLong("updatedAt"));
        } catch (SQLException ignored) {
            state.setUpdatedAt(0L);
        }
        state.setSource(nullSafeString(resultSet, "source"));
        state.setSeriesCategorySnapshot(nullSafeString(resultSet, "seriesCategorySnapshot"));
        state.setSeriesChannelSnapshot(nullSafeString(resultSet, "seriesChannelSnapshot"));
        state.setSeriesEpisodeSnapshot(nullSafeString(resultSet, "seriesEpisodeSnapshot"));
        return state;
    }
}

File: server/src/main/kotlin/com/uiptv/db/SeriesWatchStateDb.kt
package com.uiptv.db

import com.uiptv.model.SeriesWatchState
import org.jetbrains.exposed.sql.ResultRow
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.SortOrder
import org.jetbrains.exposed.sql.Table
import org.jetbrains.exposed.sql.and
import org.jetbrains.exposed.sql.deleteWhere
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.transactions.transaction
import org.jetbrains.exposed.sql.update

object SeriesWatchStateDb {
    private const val MODE_SERIES = "series"

    @JvmStatic
    fun get(): SeriesWatchStateDb = this

    fun getBySeries(accountId: String, categoryId: String, seriesId: String): SeriesWatchState? =
        transaction(SqlConnectionRuntime.database()) {
            SeriesWatchStateTable.selectAll()
                .where {
                    (SeriesWatchStateTable.accountId eq accountId) and
                        (SeriesWatchStateTable.mode eq MODE_SERIES) and
                        (SeriesWatchStateTable.categoryId eq categoryId) and
                        (SeriesWatchStateTable.seriesId eq seriesId)
                }
                .limit(1)
                .firstOrNull()
                ?.toSeriesWatchState()
        }

    fun getBySeries(accountId: String, seriesId: String): List<SeriesWatchState> =
        transaction(SqlConnectionRuntime.database()) {
            SeriesWatchStateTable.selectAll()
                .where {
                    (SeriesWatchStateTable.accountId eq accountId) and
                        (SeriesWatchStateTable.mode eq MODE_SERIES) and
                        (SeriesWatchStateTable.seriesId eq seriesId)
                }
                .orderBy(SeriesWatchStateTable.id to SortOrder.ASC)
                .map(ResultRow::toSeriesWatchState)
        }

    fun getByAccount(accountId: String, categoryId: String): List<SeriesWatchState> =
        transaction(SqlConnectionRuntime.database()) {
            SeriesWatchStateTable.selectAll()
                .where {
                    (SeriesWatchStateTable.accountId eq accountId) and
                        (SeriesWatchStateTable.mode eq MODE_SERIES) and
                        (SeriesWatchStateTable.categoryId eq categoryId)
                }
                .orderBy(SeriesWatchStateTable.id to SortOrder.ASC)
                .map(ResultRow::toSeriesWatchState)
        }

    fun getByAccount(accountId: String): List<SeriesWatchState> =
        transaction(SqlConnectionRuntime.database()) {
            SeriesWatchStateTable.selectAll()
                .where {
                    (SeriesWatchStateTable.accountId eq accountId) and
                        (SeriesWatchStateTable.mode eq MODE_SERIES)
                }
                .orderBy(SeriesWatchStateTable.id to SortOrder.ASC)
                .map(ResultRow::toSeriesWatchState)
        }

    fun upsert(state: SeriesWatchState) {
        transaction(SqlConnectionRuntime.database()) {
            val updated = SeriesWatchStateTable.update({
                (SeriesWatchStateTable.accountId eq state.accountId) and
                    (SeriesWatchStateTable.mode eq state.mode) and
                    (SeriesWatchStateTable.categoryId eq state.categoryId) and
                    (SeriesWatchStateTable.seriesId eq state.seriesId)
            }) { row ->
                row[episodeId] = state.episodeId
                row[episodeName] = state.episodeName
                row[season] = state.season
                row[episodeNum] = state.episodeNum
                row[updatedAt] = state.updatedAt
                row[sourceValue] = state.source
                row[seriesCategorySnapshot] = state.seriesCategorySnapshot
                row[seriesChannelSnapshot] = state.seriesChannelSnapshot
                row[seriesEpisodeSnapshot] = state.seriesEpisodeSnapshot
            }
            if (updated == 0) {
                SeriesWatchStateTable.insert { row ->
                    row[accountId] = state.accountId
                    row[mode] = state.mode
                    row[categoryId] = state.categoryId
                    row[seriesId] = state.seriesId
                    row[episodeId] = state.episodeId
                    row[episodeName] = state.episodeName
                    row[season] = state.season
                    row[episodeNum] = state.episodeNum
                    row[updatedAt] = state.updatedAt
                    row[sourceValue] = state.source
                    row[seriesCategorySnapshot] = state.seriesCategorySnapshot
                    row[seriesChannelSnapshot] = state.seriesChannelSnapshot
                    row[seriesEpisodeSnapshot] = state.seriesEpisodeSnapshot
                }
            }
        }
    }

    fun clear(accountId: String, categoryId: String, seriesId: String) {
        transaction(SqlConnectionRuntime.database()) {
            SeriesWatchStateTable.deleteWhere {
                (this.accountId eq accountId) and
                    (mode eq MODE_SERIES) and
                    (this.categoryId eq categoryId) and
                    (this.seriesId eq seriesId)
            }
        }
    }

    fun deleteByAccount(accountId: String) {
        transaction(SqlConnectionRuntime.database()) {
            SeriesWatchStateTable.deleteWhere { this.accountId eq accountId }
        }
    }

    fun clearAllSeries() {
        transaction(SqlConnectionRuntime.database()) {
            SeriesWatchStateTable.deleteWhere { SeriesWatchStateTable.mode eq MODE_SERIES }
        }
    }
}

private object SeriesWatchStateTable : Table(DatabaseUtils.DbTable.SERIES_WATCH_STATE_TABLE.tableName) {
    val id = integer("id").autoIncrement()
    val accountId = text("accountId").nullable()
    val mode = text("mode").nullable()
    val categoryId = text("categoryId").nullable()
    val seriesId = text("seriesId").nullable()
    val episodeId = text("episodeId").nullable()
    val episodeName = text("episodeName").nullable()
    val season = text("season").nullable()
    val episodeNum = integer("episodeNum").nullable()
    val updatedAt = long("updatedAt").nullable()
    val sourceValue = text("source").nullable()
    val seriesCategorySnapshot = text("seriesCategorySnapshot").nullable()
    val seriesChannelSnapshot = text("seriesChannelSnapshot").nullable()
    val seriesEpisodeSnapshot = text("seriesEpisodeSnapshot").nullable()

    override val primaryKey = PrimaryKey(id)
}

private fun ResultRow.toSeriesWatchState(): SeriesWatchState =
    SeriesWatchState(
        dbId = this[SeriesWatchStateTable.id].toString(),
        accountId = this[SeriesWatchStateTable.accountId],
        mode = this[SeriesWatchStateTable.mode],
        categoryId = this[SeriesWatchStateTable.categoryId],
        seriesId = this[SeriesWatchStateTable.seriesId],
        episodeId = this[SeriesWatchStateTable.episodeId],
        episodeName = this[SeriesWatchStateTable.episodeName],
        season = this[SeriesWatchStateTable.season],
        episodeNum = this[SeriesWatchStateTable.episodeNum] ?: 0,
        updatedAt = this[SeriesWatchStateTable.updatedAt] ?: 0L,
        source = this[SeriesWatchStateTable.sourceValue],
        seriesCategorySnapshot = this[SeriesWatchStateTable.seriesCategorySnapshot],
        seriesChannelSnapshot = this[SeriesWatchStateTable.seriesChannelSnapshot],
        seriesEpisodeSnapshot = this[SeriesWatchStateTable.seriesEpisodeSnapshot]
    )
